service/history/replication/eager_namespace_refresher.go

```diff
@@ -118,3 +144,44 @@ func (e *eagerNamespaceRefresherImpl) UpdateNamespaceFailoverVersion(namespaceId
 	}
 	return nil
 }
+
+func (e *eagerNamespaceRefresherImpl) SyncNamespaceFromSourceCluster(ctx context.Context, namespaceId namespace.ID, sourceCluster string) error {
+	// Potential Perf improvement: one lock per namespaceID
+	e.lock.Lock()
```

Comment:
why do we need this lock?

Reply:
Just in case multiple caller is trying invoke this method at same time for the same namespace. i.e.
This API will be invoked by passive side when passive side received Replication task AND passive side does not have the namespace info. It's very possible that multiple replication tasks for the same namespace arrive in a very short period of time. So use a simple lock here to prevent calling the remote cluster multiple times.